accepted/2021/runtime-security-mitigations.md

```diff
@@ -0,0 +1,213 @@
+
+## ARM Pointer Authentication (PA)
+
+A significant number of address bits on 64-bit systems is unused. [Pointer authentication](https://www.qualcomm.com/media/documents/files/whitepaper-pointer-authentication-on-armv8-3.pdf) uses the unused address bits to store pointer authentication code (PAC), computed as a hash of pointer value, secret key and pointer type specific context. The PAC is validated before each pointer use. The scheme is supported by new instructions to compute and validate PAC.
+
+PA can be used either on code pointers for control flow integrity checks or on data pointers for general data sanity checks.
+
+PA in compilation with BTI provides stronger protection than IBT for indirect calls and branches. However, PA is weaker than shadow stack for protecting return addresses since it does not prevent pointer replacement exploits.
+
+Apple published [preview implementation](https://developer.apple.com/documentation/security/preparing_your_app_to_work_with_pointer_authentication) of ARM64E ABI hardened using pointer authentication. We expect that other major operating systems will follow with similar technology for ARM platforms. Updating .NET runtimes to be compatible and take advantage of pointer authentication is major undertaking that affects number of sub-systems, including JIT, GC, interop or debugger. Our preliminary plan is to start tackling it for .NET 7.
+
+## Stack Buffer Overflow Protection
+
+[Stack Buffer Overflow](https://en.wikipedia.org/wiki/Stack_buffer_overflow) protection is among the older security mitigations. It has been known as &quot;GS cookie&quot; on Windows. It is designed to catch writes into stack memory outside the intended bounds.
+
+The first .NET runtime that included it was .NET Framework 2.0, and the implementation has not changed much since then. The mitigation in C/C++ compilers have seen improvements in the meantime (for example, [Enhanced GS](https://msrc-blog.microsoft.com/2009/03/20/enhanced-gs-in-visual-studio-2010/) or -[fstack-protector-strong](https://lwn.net/Articles/584225/)).
+
+We will review and update stack buffer overflow protection in .NET runtime to match state of the art.
+
+# Language, Libraries and Runtime Features
+
+Library and language design have always played a key role in mitigating security bugs by preventing or discouraging them from being introduced in the first place.
+
+## Unsafe API and C# language constructs
+
+Number of newer APIs such as `Memory<T>` and friends come with security risks that are not clearly communicated.
+
+We need to adopt a more holistic approach to unsafe features and APIs that we expose for public use. Developers need feedback in the IDE so that they understand that they are using constructs that are unsafe and are [informed of the specific risks that come with using those APIs](https://github.com/dotnet/runtime/issues/41418). Over time, we bias more to constructs and paradigms that are safe by default (possibly inspired by Rust).
```

Comment:
Opinion: An IDE warning is not enough. Not everyone will use Visual Studio. There needs to be non-IDE feedback, even if it's just compile time warnings.